core/src/main/java/org/jsmart/zerocode/core/runner/ZeroCodeMultiStepsScenarioRunnerImpl.java

```diff
@@ -102,8 +102,19 @@ public synchronized boolean runScenario(ScenarioSpec scenario, RunNotifier notif
 
             for (Step thisStep : scenario.getSteps()) {
 
-                final int stepLoopTimes = thisStep.getLoop() == null ? 1 : thisStep.getLoop();
+                boolean retryTillSuccess = false;
+                int delay = 0;
+                int stepLoopTimes = 1;
+                if ( thisStep.getLoop() != null ) {
+                    stepLoopTimes = thisStep.getLoop();
+                } else if ( thisStep.getRetry() != null ) {
+                    stepLoopTimes = thisStep.getRetry().getMax();
+                    delay = thisStep.getRetry().getDelay();
+                    retryTillSuccess = true;
+                }
+
                 for (int i = 0; i < stepLoopTimes; i++) {
+                    waitForDelay(delay);
```

Comment:
The delay should only be  used for retires(incase of failures), not to introduce delays for iteration of every loop

Reply:
The primary scenario I was considering was checking the result of an asynchronous operation, where an initial delay would be sensible.
But maybe that should be expressed outside of the retry-structure, in an 'initial-delay' attribute of the Step?
OK, I'll adapt the retry-delay.